### PR TITLE
TVM support in APPLib added

### PR DESCRIPTION
**Pull request by ANSHUMAN87**
TVM support in APPLib added

### CODE REVIEW

**Comment by sharathks84** [comment]
- .idea folder to be deleted
- also TVM changes let Vinod review this once.

**Comment by ANSHUMAN87** [comment]
>     * .idea folder to be deleted
> 
>     * also TVM changes let Vinod review this once.

@sharathks84: Thanks! .idea folder is deleted now.

File: tvm/src/main/java/org/apache/tvm/Function.java [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+  Function(long handle, boolean isResident) {
+    super(ArgTypeCode.FUNC_HANDLE);
+    this.handle = handle;
+    this.isResident = isResident;
+  }
+
+  Function(long handle) {
+    this(handle, false);
+  }
+
```
Comment (VinodSKumar) [inline tvm/src/main/java/org/apache/tvm/Function.java:91]: Any reason why finalize has been deleted, how is the resources released?

File: tvm/src/main/java/org/apache/tvm/Module.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+   * @return The loaded module
+   */
+  public static Module load(String path, String fmt) {
+    TVMValue ret = getApi("ModuleLoadFromFile").pushArg(path).pushArg(fmt).invoke();
+    assert ret.typeCode == ArgTypeCode.MODULE_HANDLE;
+    return ret.asModule();
+  }
+
+  public static Module load(String path) {
+    LogUtil.info("Jai Hanuman!!!", path);
```
Comment (VinodSKumar) [inline tvm/src/main/java/org/apache/tvm/Module.java]: Please remove dev debug logs, if required update with appropriate logs. Please take care of this in all locations.

File: tvm/src/main/java/org/apache/tvm/NativeLibraryLoader.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+      return WINDOWS;
+    }
+    return osname;
+  }
+
+  private static String getUnifiedOSName() {
+    return unifyOSName(System.getProperty("os.name"));
+  }
+
+  private static File createTempFile(String name) throws NullPointerException {
```
Comment (VinodSKumar) [inline tvm/src/main/java/org/apache/tvm/NativeLibraryLoader.java]: It needs to be io exception, and null pointer exception could be avoided using a check

File: tvm/src/main/java/org/apache/tvm/NDArrayBase.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+   * <p>
+   * We highly recommend you to do this manually since the GC strategy is lazy.
+   * </p>
+   */
+  public void release() {
+    if ((!isReleased) && (!isView)) {
+      Base.checkCall(Base._LIB.tvmArrayFree(handle));
+      isReleased = true;
+    }
+  }
```
Comment (VinodSKumar) [inline tvm/src/main/java/org/apache/tvm/NDArrayBase.java:68]: finalize has been deleted, I think it is required to release resources

File: tvm/src/main/java/org/apache/tvm/NDArray.java [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+  private final TVMType dtype;
+  private final Device device;
+  private static final String ARRAY = " array";
+
+  NDArray(long handle, boolean isView, TVMType dtype, Device dev) {
+    super(handle, isView);
+    this.dtype = dtype;
+    this.device = dev;
+  }
+
```
Comment (VinodSKumar) [inline tvm/src/main/java/org/apache/tvm/NDArray.java:38]: Finalize has been removed, please check this for all other classes also

File: tvm/src/main/java/org/apache/tvm/TVMType.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  public TVMType(String typeStr) {
+    this(typeStr, 1);
+  }
+
+  @Override public int hashCode() {
+    return (typeCode << 16) | (bits  << 8) | lanes;
+  }
+
+  @Override public boolean equals(Object other) {
+    if (other instanceof TVMType) {
```
Comment (VinodSKumar) [inline tvm/src/main/java/org/apache/tvm/TVMType.java:70]: Why is null check removed

File: tvm/src/main/jni/runtime/dso_library.cc [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+ * \brief Create library module to load from dynamic shared library.
+ */
+#include <tvm/runtime/memory.h>
+#include <tvm/runtime/module.h>
+#include <tvm/runtime/packed_func.h>
+#include <tvm/runtime/registry.h>
+
+#include "library_module.h"
+
+#ifdef DMLC_LOG_STACK_TRACE
```
Comment (VinodSKumar) [inline tvm/src/main/jni/runtime/dso_library.cc:32]: Why is this being done, could we add a comment as it is not straight forward to understand

File: tvm/src/main/jni/tvm/3rdparty/dmlc-core/include/dmlc/endian.h [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  #elif defined(__sun) || defined(sun)
+    #include <sys/isa_defs.h>
+    #if defined(_LITTLE_ENDIAN)
+      #define DMLC_LITTLE_ENDIAN 1
+    #else
+      #define DMLC_LITTLE_ENDIAN 0
+    #endif
+  #else
+    //#error "Unable to determine endianness of your machine; use CMake to compile"
+    #include <endian.h>
```
Comment (VinodSKumar) [inline tvm/src/main/jni/tvm/3rdparty/dmlc-core/include/dmlc/endian.h:34]: Would it be the default case, or should it be for CMAKE case?

File: tvm/src/main/jni/tvm/3rdparty/dmlc-core/include/dmlc/endian.h [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    #define DMLC_LITTLE_ENDIAN (__BYTE_ORDER == __LITTLE_ENDIAN)
```
Comment (VinodSKumar) [inline tvm/src/main/jni/tvm/3rdparty/dmlc-core/include/dmlc/endian.h:35]: Why is the little endian being hardcoded?

File: tvm/src/main/jni/tvm/3rdparty/dmlc-core/include/dmlc/logging.h [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+#define DMLC_LOGGING_H_
+#include <cstdio>
+#include <cstdlib>
+#include <string>
+#include <vector>
+#include <stdexcept>
+#include <memory>
+#include "./base.h"
+
+#ifdef DMLC_LOG_STACK_TRACE
```
Comment (VinodSKumar) [inline tvm/src/main/jni/tvm/3rdparty/dmlc-core/include/dmlc/logging.h:19]: Same as before a comment would be required here

**Review by VinodSKumar — COMMENTED** [review]
We can add a comment to mention that RPC has been removed, but in future once HMOS supports all +8 and N devices, it is a good use case for RPC. Later it can be added.

File: tvm/src/main/java/org/apache/tvm/Function.java (hunk 1 above)
Comment (ANSHUMAN87) [reply tvm/src/main/java/org/apache/tvm/Function.java:91]: The typical and recommended usage would be as below:
Function api = Function.getFunction("API name");
api.release();

File: tvm/src/main/java/org/apache/tvm/NativeLibraryLoader.java (hunk 3 above)
Comment (ANSHUMAN87) [reply tvm/src/main/java/org/apache/tvm/NativeLibraryLoader.java]: NativeLibrary is an internal implementation of TVM, user wont be using these APIs, so IMO it wont make any difference being NullPointerException. Please let me know your opinion. TIA!

File: tvm/src/main/java/org/apache/tvm/NDArrayBase.java (hunk 4 above)
Comment (ANSHUMAN87) [reply tvm/src/main/java/org/apache/tvm/NDArrayBase.java:68]: Same as in my last reply in Function, even here also it is highly recommended to use release() method.

File: tvm/src/main/java/org/apache/tvm/NDArray.java (hunk 5 above)
Comment (ANSHUMAN87) [reply tvm/src/main/java/org/apache/tvm/NDArray.java:38]: same as above

File: tvm/src/main/java/org/apache/tvm/TVMType.java (hunk 6 above)
Comment (ANSHUMAN87) [reply tvm/src/main/java/org/apache/tvm/TVMType.java:70]: instanceof handles null check too

File: tvm/src/main/jni/runtime/dso_library.cc (hunk 7 above)
Comment (ANSHUMAN87) [reply tvm/src/main/jni/runtime/dso_library.cc:32]: DMLC Log stack is not required for us, as HMOS has its own logging framework. Will add comment

File: tvm/src/main/jni/tvm/3rdparty/dmlc-core/include/dmlc/endian.h (hunk 9 above)
Comment (ANSHUMAN87) [reply tvm/src/main/jni/tvm/3rdparty/dmlc-core/include/dmlc/endian.h:35]: This is same as case for all the flavors as can be refereed in above code as well.As the code are being compiled in windows machine, but target is for arm, the cmake detection logic has some issue. So this default case was added.

File: tvm/src/main/jni/tvm/3rdparty/dmlc-core/include/dmlc/endian.h (hunk 8 above)
Comment (ANSHUMAN87) [reply tvm/src/main/jni/tvm/3rdparty/dmlc-core/include/dmlc/endian.h:34]: It is a default case. I will remove the commented line, that is confusing i understand.

File: tvm/src/main/java/org/apache/tvm/NativeLibraryLoader.java (hunk 3 above)
Comment (VinodSKumar) [reply tvm/src/main/java/org/apache/tvm/NativeLibraryLoader.java]: Semantically it is not correct, better to have it as IO exception

Comment (ANSHUMAN87) [reply tvm/src/main/java/org/apache/tvm/NativeLibraryLoader.java]: Done!

**Review by VinodSKumar — APPROVED** [review]
OK

**Review by sharathks84 — APPROVED** [review]
OKAY

**Review by Gleelakrishna — APPROVED** [review]

**Pull request merged**